core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -349,9 +351,13 @@ public synchronized IndexShard createShard(ShardRouting routing) throws IOExcept
             shards = newMapBuilder(shards).put(shardId.id(), indexShard).immutableMap();
             success = true;
             return indexShard;
+        } catch (ShardLockObtainFailedException e) {
+            throw new IOException("failed to obtain in-memory shard lock", e);
         } finally {
             if (success == false) {
-                IOUtils.closeWhileHandlingException(lock);
+                if (lock != null) {
+                    IOUtils.closeWhileHandlingException(lock);
```

Comment:
Believe it or not, it's fine to pass `null` to `IOUtils.closeWhileHandlingException` (it just skips them).  We do that for cases where you might have N things that need closing from a large try block, and you don't know which are `null` and which are not (depends on where the exception was thrown).

But for here I like the `null` check: less smelly.